lib/resque/durable/background_heartbeat.rb

```diff
@@ -0,0 +1,83 @@
+require 'thread'
+
+module Resque
+  module Durable
+
+    # Creates a background thread to regularly heartbeat the queue audit.
+    class BackgroundHeartbeat
+      DEFAULT_INTERVAL = 15
+
+      def initialize(queue_audit, interval = nil)
+        @queue_audit  = queue_audit
+        @last_timeout = nil
+        @interval     = interval || DEFAULT_INTERVAL
+        @mutex        = Mutex.new
+        @cv           = ConditionVariable.new
+        @stop         = false
+        @thread       = nil
+      end
+
+      class << self
+        # only a separate method for easy stubbing
+        def exit_now!
+          abort
+        end
+      end
+
+      def with_heartbeat
+        start!
+        yield
+      ensure
+        stop_and_wait!
+      end
+
+      def heartbeat!
+        @last_timeout ||= @queue_audit.timeout_at
+        @last_timeout = @queue_audit.optimistic_heartbeat!(@last_timeout)
+      rescue StandardError => e
+        @queue_audit.logger.error("Exception in BackgroundHeartbeat thread: #{e.class.name}: #{e.message}")
+        self.class.exit_now!
+      ensure
+        ActiveRecord::Base.clear_active_connections!
+      end
+
+      def start!
+        raise "Thread is already running!" if @thread
+        @stop = false
+
+        # Perform immediately to reduce heartbeat race condition opportunities
+        heartbeat!
+
+        @thread = Thread.new do
+          while !@stop
+            heartbeat!
+
+            @mutex.synchronize do
+              @cv.wait(@mutex, @interval)
+            end
```

Comment:
As far as I understand the mutex sleep correctly, I think you can actually get by with a `@mutex.sleep` here and a `@thread.wakeup` in signal_stop! instead of using ConditionVariable since you know each object explicitly.